Better docs for fetch event (#27367)
mdn/content: files/en-us/web/api/serviceworkerglobalscope/fetch_event/index.md

onfetch = (event) => {};
```

## Description

> The `fetch` event is fired in the service worker's global scope when the main app thread makes a network request. This includes not only explicit {{domxref("fetch()")}} calls from the main thread, but implicit network requests made when pages and subresources such as JavaScript, CSS, and images are loaded.

Author: ```suggestion
The `fetch` event is fired in the service worker's global scope when the main app thread makes a network request. This includes not only explicit {{domxref("fetch()")}} calls from the main thread, but also implicit network requests made when pages and subresources such as JavaScript, CSS, and images are loaded.
```

> The `fetch` event is fired in the service worker's global scope when the main app thread makes a network request. This includes not only explicit {{domxref("fetch()")}} calls from the main thread, but also implicit network requests made when pages and subresources such as JavaScript, CSS, and images are loaded.

Reviewer: How about

```suggestion
The `fetch` event is fired in the service worker's global scope when the main app thread makes a network request. This includes not only explicit {{domxref("fetch()")}} calls from the main thread, but also implicit network requests to load pages and subresources (such as JavaScript, CSS, and images) made by the browser following page navigation.
```

> The `fetch` event is fired in the service worker's global scope when the main app thread makes a network request. This includes not only explicit {{domxref("fetch()")}} calls from the main thread, but also implicit network requests made when pages and subresources such as JavaScript, CSS, and images are loaded.

Author: Hm, do you have a suggestion here? I want to emphasise that a single "page load" is typically several network requests. Maybe "but also implicit network requests on navigation for pages and their subresources (such as JavaScript, CSS, and images)." ? That seems clumsy to me though. Or "but also implicit network requests that are made when the user navigates to new pages, for the main document but also for subresources such as JavaScript, CSS, and images." ?

> The `fetch` event is fired in the service worker's global scope when the main app thread makes a network request. This includes not only explicit {{domxref("fetch()")}} calls from the main thread, but also implicit network requests made when pages and subresources such as JavaScript, CSS, and images are loaded.

Reviewer: @wbamberg Some variant of my suggestion above? ^^^

The event handler is passed a {{domxref("FetchEvent")}} object, which provides access to the request as a {{domxref("Request")}} instance.

> The `FetchEvent` also provides a {{domxref("FetchEvent.respondWith()", "respondWith()")}} method, that takes a {{domxref("Response")}} (or a `Promise` that resolves to a `Response`) as a parameter, and this enables the event handler to return a different response to the request: for example, it enables a service worker to return:

Reviewer: ```suggestion
The `FetchEvent` also provides a {{domxref("FetchEvent.respondWith()", "respondWith()")}} method, that takes a {{domxref("Response")}} (or a `Promise` that resolves to a `Response`) as a parameter.
This enables the event handler to return a different response to the request: for example, it enables a service worker to return:
```

- A locally cached response retrieved from the {{domxref("Cache")}} interface.
- A response that the service worker synthesizes, using methods like {{domxref("Response.json()")}} or the {{domxref("Response.Response()", "Response()")}} constructor.
- A network error, using the {{domxref("Response.error_static()", "Response.error()")}} method. This will cause the `fetch()` call to reject.

> The `respondWith()` method can only be called once for a given request. If multiple `fetch` event listeners are added, they will be called in the order they were registered until one of them calls `respondWith()`.

Reviewer 2: Yes interesting

The `respondWith()` method must be called synchronously: that is, you can't call it in a `then` handler.

Typically, a `fetch` event handler will execute different strategies depending on features of the request such as its URL:

```js
function strategy1() {
  return fetch("picnic.jpg");
}

function strategy2() {
  return Response.error();
}

const pattern1 = /^\/salamander/;
const pattern2 = /^\/lizard/;

self.addEventListener("fetch", (event) => {
  const url = new URL(event.request.url);
  if (pattern1.test(url.pathname)) {
    event.respondWith(strategy1());
  } else if (pattern2.test(url.pathname)) {
    event.respondWith(strategy2());
  }
});
```

> If `respondWith()` is not called in the handler, then the original network request is made: so in the example above, all requests that do not match `pattern1` or `pattern2` are made as if the service worker did not exist.

Reviewer: ```suggestion
If `respondWith()` is not called in the handler, then the user-agent automatically makes the original network request: so in the example above, all requests that do not match `pattern1` or `pattern2` are made as if the service worker did not exist.
```

## Event type

A {{domxref("FetchEvent")}}.

## Examples

### Cache falling back to network

This `fetch` event handler first tries to find the response in the cache. If a response is found, it returns the cached response. Otherwise, it tries to fetch the resource from the network.

```js
async function cacheThenNetwork(request) {
  const cachedResponse = await caches.match(request);
  if (cachedResponse) {
    console.log("Found response in cache:", cachedResponse);
    return cachedResponse;
  }
  console.log("Falling back to network");
  return fetch(request);
}

self.addEventListener("fetch", (event) => {
  console.log(`Handling fetch event for ${event.request.url}`);
  event.respondWith(cacheThenNetwork(event.request));
});
```

> 

Reviewer: Should we have a note here explaining that we didn't have to explicitly call respondWith for the network case? I mean we already covered that above, but is it worth explaining that this particular example choses to make the request anyway.

### Cache only

> This `fetch` event handler implements a "cache only" policy for scripts and stylesheets. If the request's {{domxref("Request.destination", "destination")}} property is `"script"` or `"style"`, the handler only looks in the cache, returning an error if the response was not found.
> 
> All other requests go through to the network.

Reviewer: ```suggestion
This `fetch` event handler implements a "cache only" policy for scripts and stylesheets. If the request's {{domxref("Request.destination", "destination")}} property is `"script"` or `"style"`, the handler only looks in the cache, returning an error if the response was not found.
All other requests go through to the network.
```
